File: backend/schemas/leads.py
"""Lead acquisition HTTP contracts."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


RunMode = Literal["FULL_SCAN", "INCREMENTAL", "REFRESH_STALE", "ENRICH_ONLY", "VERIFY_ONLY"]


class LeadRunRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    mode: RunMode = "FULL_SCAN"
    sources: list[str] = Field(default_factory=list)
    governorates: list[str] = Field(default_factory=list)
    segments: list[str] = Field(default_factory=list)
    keyword_set: Literal["approved", "custom"] = "approved"
    keywords: list[str] = Field(default_factory=list, max_length=1000)
    dry_run: bool = False
    execute: bool = True
    adaptive_tiling: bool = True
    max_records_per_job: int | None = Field(default=None, ge=1, le=10000)
    source_options: dict[str, Any] = Field(default_factory=dict)


class LeadRunPlan(BaseModel):
    dry_run: bool = True
    enabled_sources: list[str]
    governorates: list[str]
    segments: list[str]
    keyword_count: int
    planned_jobs: int
    missing_credentials: list[str]
    warnings: list[str]


class LeadJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    job_uid: str
    source_uid: str
    governorate: str | None
    tile: dict[str, Any] | None
    category: str | None
    status: str
    attempt: int
    checkpoint: dict[str, Any]
    processed: int
    found: int
    unique_count: int
    duplicates: int
    error: str | None
    started_at: datetime | None
    completed_at: datetime | None


class LeadRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    run_uid: str
    name: str
    mode: str
    sources: list[str]
    geography: dict[str, Any]
    segments: list[str]
    keywords: list[str]
    status: str
    planned_jobs: int
    processed: int
    found: int
    unique_count: int
    duplicates: int
    errors: int
    current_source: str | None
    current_governorate: str | None
    current_category: str | None
    warnings: list[str]
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    jobs: list[LeadJobResponse] = Field(default_factory=list)


class LeadExportRequest(BaseModel):
    format: Literal["csv", "xlsx", "parquet"]
    fit_class: str | None = None
    governorate: str | None = None
    category: str | None = None
    source: str | None = None
    verified_only: bool = False

File: frontend/src/types/leads.ts
export type LeadSource = {
  source_uid: string; name: string; type: string; enabled: boolean; configured: boolean;
  bulk_support: boolean; query_support: boolean; geo_support: string; credentials_required: boolean;
  freshness: string; last_run?: string; last_success?: string; terms_status: string;
  storage_policy: string; rate_limit: string; status: string; status_detail?: string;
};

export type LeadControl = {
  workbook: { available: boolean; status: string; active_command_source: string; placement: string; filename?: string; sheet_counts: Record<string, number>; query_matrix_rows: number; enabled_query_jobs: number; errors: string[] };
  active_command_source: string;
  query_matrix_rows: number;
  enabled_query_jobs: number;
  segments: { category_id: string; name: string; tier: number; buyer_type: string; terms: string[] }[];
  keywords: { set: string; count: number; languages: string[] };
  governorates: { id: string; name: string; name_ar: string; bbox: number[]; density: string }[];
  country: { code: string; name: string; bbox: number[]; polygon: number[][] };
  run_modes: string[];
};

export type LeadRunRequest = {
  name: string; mode: string; sources: string[]; governorates: string[]; segments: string[];
  keyword_set: "approved" | "custom"; keywords: string[]; dry_run: boolean; execute: boolean;
  adaptive_tiling: boolean; max_records_per_job?: number; source_options: Record<string, unknown>;
};

export type LeadRunPlan = {
  dry_run: true; enabled_sources: string[]; governorates: string[]; segments: string[];
  keyword_count: number; planned_jobs: number; missing_credentials: string[]; warnings: string[];
};

export type LeadJob = {
  job_uid: string; source_uid: string; governorate?: string; category?: string; status: string;
  processed: number; found: number; unique_count: number; duplicates: number; error?: string;
};

export type LeadRun = {
  dry_run: false; run_uid: string; name: string; mode: string; sources: string[]; geography: Record<string, unknown>;
  segments: string[]; keywords: string[]; status: string; planned_jobs: number; processed: number;
  found: number; unique_count: number; duplicates: number; errors: number; current_source?: string;
  current_governorate?: string; current_category?: string; warnings: string[]; progress_percent: number;
  started_at?: string; completed_at?: string; created_at: string; jobs?: LeadJob[];
};

export type LeadSummary = {
  lead_uid: string; business: string; category?: string; fit_class: string; governorate?: string;
  city?: string; phone?: string; website?: string; source: string; score: number;
  freshness: string; verification: string;
};

export type Page<T> = { items: T[]; total: number; offset: number; limit: number };
export type LeadStats = { total_leads: number; classes: Record<"A+" | "A" | "B" | "C", number>; active_runs: number; sources: number; last_run?: string; last_run_status?: string; errors: number };

